### PR TITLE
Latest GCC (8) & fix for cse (fixes #1463)

### LINKED ISSUE #1463: std::set_intersection called with unsorted ranges
A test suite in a library of mine which uses symengine is failing when using `cse` (I caught this bug when compiling using gcc with `_DGLIBCXX_DEBUG` defined).

I've tracked down the error to this line:
https://github.com/symengine/symengine/blob/b3a2159b7e6950726157626069c30712427d55b2/symengine/cse.cpp#L147

The test case is complicated and the library depends on many other 3rd party libraries and I have yet to write a minimal test case for symengine.

<details><summary>Click here to show an excerpt from my 3rd party test log:</summary>
<pre>
/usr/include/c++/8/bits/stl_algo.h:5269:
Error: elements in iterator range [__first1, __last1) are not sorted.

Objects involved in the operation:
    iterator "__first1" @ 0x0x7fffee840910 {
      type = __gnu_debug::_Safe_iterator<__gnu_cxx::__normal_iterator<unsigned int*, std::__cxx1998::vector<unsigned int, std::allocator<unsigned int> > >, std::__debug::vector<unsigned int, std::allocator<unsigned int> > > (mutable iterator);
      state = dereferenceable (start-of-sequence);
      references sequence with type 'std::__debug::vector<unsigned int, std::allocator<unsigned int> >' @ 0x0x7fffee840c90
    }
    iterator "__last1" @ 0x0x7fffee840940 {
      type = __gnu_debug::_Safe_iterator<__gnu_cxx::__normal_iterator<unsigned int*, std::__cxx1998::vector<unsigned int, std::allocator<unsigned int> > >, std::__debug::vector<unsigned int, std::allocator<unsigned int> > > (mutable iterator);
      state = past-the-end;
      references sequence with type 'std::__debug::vector<unsigned int, std::allocator<unsigned int> >' @ 0x0x7fffee840c90
    }
</pre>
</details>

### PR DESCRIPTION
**Pull request by bjodah**
While #1465 is functional, in this PR I'll try to make Travis *fail* for the regression test, before applying the fix.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -69,7 +69,7 @@ matrix:
           - ubuntu-toolchain-r-test
           packages:
           - binutils-dev
-          - g++-5
+          - g++-8
```
Comment (isuruf) [inline .travis.yml]: You'll have to fix https://github.com/symengine/symengine/blob/master/bin/install_travis.sh#L35 as well

**Comment by bjodah** [comment]
The build with latest gcc uses `-Werror`, gcc-8 has introduced some new warnings. Should I drop `-Werror` from that build for now? (and add it to another build using an older gcc version?)

**Comment by bjodah** [comment]
Next warning to fail:
```
/home/travis/build/symengine/symengine/symengine/tests/basic/test_relationals.cpp: In function ‘void ____C_A_T_C_H____T_E_S_T____10()’:
/home/travis/build/symengine/symengine/symengine/tests/basic/test_relationals.cpp:178:43: error: catching polymorphic type ‘class SymEngine::SymEngineException’ by value [-Werror=catch-value=]
     CHECK_THROWS_AS(Lt(ComplexInf, zero), SymEngineException);
                                           ^~~~~~~~~~~~~~~~~~
```

**Comment by bjodah** [comment]
Success, the test triggers [the failure](https://travis-ci.org/symengine/symengine/jobs/444168824#L1520):
```
/usr/include/c++/8/bits/stl_algo.h:5269:
Error: elements in iterator range [__first1, __last1) are not sorted.
```

**Comment by bjodah** [comment]
gcov seem to be failing:
```
==> Running gcov in /home/travis/build/symengine/symengine (disable via -X gcov)
/home/travis/build/symengine/symengine/build/symengine/CMakeFiles/symengine.dir/parser/parser.cpp.gcno:'_ZN7Teuchos14TypeNameTraitsINS_11RCPNodeTmplIN9SymEngine8RealMPFRENS_13DeallocDeleteIS3_EEEEE12concreteNameERKS6_' has arcs to entry block
```
is it cache related?

**Comment by isuruf** [comment]
I reset the cache before the last job. (ccache does track the gcc version so that this shouldn't be an issue at all.)

**Comment by bjodah** [comment]
Didn't wait for Appyveyor to finish since it past for the previous commit, and the last commit only affected gcov.

**Comment by certik** [comment]
@bjodah thanks for fixing it. For reference, why does it now require to use `NotImplementedError &` instead of `NotImplementedError`? Is C++ now somehow returning a reference to the exception? I am confused.

**Comment by isuruf** [comment]
@certik,

We can either do,
```
try
{

}
catch (NotImplementedError e)
{

}
```
or
```
try
{

}
catch (NotImplementedError& e)
{

}
```

We had the first one, but we are catching a polymorphic type by value, (A type with a virtual function) which means we only have memory allocated for `NotImplementedError` type. If `e` was an object of a child class of `NotImplementedError` which had more data members than `NotImplementedError`, then this leads to crashes. (Or at least undefined behaviour)

**Comment by bjodah** [comment]
I simply followed @isurf's fix and extended it to the other exception types, since it's a macro I assumed it was simply used as a type specifier in the generated code. Looking at the implementation in catch (in our checked-in version) I see:
```c++
#define CHECK_THROWS_AS( expr, exceptionType ) INTERNAL_CATCH_THROWS_AS( "CHECK_THROWS_AS", exceptionType, Catch::ResultDisposition::ContinueOnFailure, expr )
```
and
```c++
#define INTERNAL_CATCH_THROWS_AS( macroName, exceptionType, resultDisposition, expr ) \
    do { \
        Catch::ResultBuilder __catchResult( macroName, CATCH_INTERNAL_LINEINFO, CATCH_INTERNAL_STRINGIFY(expr) ", " CATCH_INTERNAL_STRINGIFY(exceptionType), resultDisposition ); \
        if( __catchResult.allowThrows() ) \
            try { \
                static_cast<void>(expr); \
                __catchResult.captureResult( Catch::ResultWas::DidntThrowException ); \
            } \
            catch( exceptionType ) { \
                __catchResult.captureResult( Catch::ResultWas::Ok ); \
            } \
            catch( ... ) { \
                __catchResult.useActiveException( resultDisposition ); \
            } \
        else \
            __catchResult.captureResult( Catch::ResultWas::Ok ); \
        INTERNAL_CATCH_REACT( __catchResult ) \
    } while( Catch::alwaysFalse() )
```
so I guess it now catches the exception by reference

**Comment by certik** [comment]
I see. Yes, we have to use a reference for a polymorphic type. This looks almost like a bug in Catch itself, since it would be more logical (I would think) to simply write `CHECK_THROWS_AS(c2->div(*c1), NotImplementedError);` and the macro would add the `&` in there properly, since it doesn't make sense to call it without it. But perhaps some people prefer to use Catch without the `&` for non-polymorphic types.

**Pull request merged**